src/contrib/persistence/Akka.Persistence.Query.Sql/QueryThrottler.cs

```diff
@@ -0,0 +1,120 @@
+﻿// //-----------------------------------------------------------------------
+// // <copyright file="QueryThrottler.cs" company="Akka.NET Project">
+// //     Copyright (C) 2009-2023 Lightbend Inc. <http://www.lightbend.com>
+// //     Copyright (C) 2013-2023 .NET Foundation <https://github.com/akkadotnet/akka.net>
+// // </copyright>
+// //-----------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using Akka.Actor;
+using Akka.Event;
+using Akka.Pattern;
+
+namespace Akka.Persistence.Query.Sql;
+
+/// <summary>
+/// Request token from throttler
+/// </summary>
+internal sealed class RequestQueryStart
+{
+    public static readonly RequestQueryStart Instance = new RequestQueryStart();
+    private RequestQueryStart() { }
+}
+
+/// <summary>
+/// Token request granted
+/// </summary>
+internal sealed class QueryStartGranted
+{
+    public static readonly QueryStartGranted Instance = new QueryStartGranted();
+    private QueryStartGranted() { }
+}
+
+/// <summary>
+/// Return token to throttler
+/// </summary>
+internal sealed class ReturnQueryStart
+{
+    public static readonly ReturnQueryStart Instance = new ReturnQueryStart();
+    private ReturnQueryStart() { }
+}
+
+/// <summary>
+/// Token bucket throttler that grants queries permissions to run each iteration
+/// </summary>
+/// <remarks>
+/// Works identically to the RecoveryPermitter built into Akka.Persistence.
+/// </remarks>
+internal sealed class QueryThrottler : ReceiveActor
+{
+    private readonly LinkedList<IActorRef> _pending = new();
+    private readonly ILoggingAdapter _log = Context.GetLogger();
+    private int _usedPermits;
+    private int _maxPendingStats;
+
+    public QueryThrottler(int maxPermits)
+    {
+        MaxPermits = maxPermits;
+        
+        Receive<RequestQueryStart>(_ =>
+        {
+            Context.Watch(Sender);
```

Comment:
Death watch requestors of tokens, so we can return used tokens in the event of dead requestors.